synth/src/datasource/relational_datasource.rs

```diff
@@ -39,123 +43,199 @@ pub struct ForeignKey {
-                        column_info.data_type, collection_name, column_info.column_name
-                    ),
-                    //TODO: More variants
-                    _ => {}
-                }
-            }
-        }
+    /// Gets a multithread pool to execute queries with
+    fn get_multithread_pool(&self) -> Pool<Self::DB>;
 
-        let column_names = first_valueset
-            .keys()
-            .cloned()
-            .collect::<Vec<String>>()
-            .join(",");
-
-        let mut futures = Vec::with_capacity(collection.len());
-
-        for rows in collection.chunks(batch_size) {
-            let mut query = format!(
-                "INSERT INTO {} ({}) VALUES \n",
-                collection_name, column_names
-            );
-
-            let mut curr_index = 0;
-            let mut query_params = vec![];
-
-            for (i, row) in rows.iter().enumerate() {
-                let row_obj = row
-                    .as_object()
-                    .expect("This is always an object (sampler contract)");
-
-                let extend = row_obj.values().len();
-                Self::extend_parameterised_query(&mut query, curr_index, extend);
-                curr_index += extend;
-                query_params.extend(row_obj.values());
-
-                if i == rows.len() - 1 {
-                    query.push_str(";\n");
-                } else {
-                    query.push_str(",\n");
-                }
-            }
-            let future = self.execute_query(query, query_params);
-            futures.push(future);
-        }
+    /// Prepare a single query with data source specifics
+    fn query<'q>(&self, query: &'q str) -> Query<'q, Self::DB, Self::Arguments> {
+        sqlx::query(query)
+    }
 
-        let results = join_all(futures).await;
+    /// Get query for table names
+    fn get_table_names_query(&self) -> &str;
 
-        if let Err(e) = results.into_iter().bcollect::<Vec<Self::QueryResult>>() {
-            bail!("One or more database inserts failed: {:?}", e)
-        }
+    /// Get query for primary keys
+    fn get_primary_keys_query(&self) -> &str;
+
+    /// Get query for foreign keys
+    fn get_foreign_keys_query(&self) -> &str;
 
-        info!("Inserted {} rows...", collection.len());
+    /// Get query for columns info
+    fn get_columns_info_query(&self) -> &str;
+
+    async fn set_seed(&self) -> Result<()> {
+        // Default for sources that don't need to set a seed
         Ok(())
     }
 
+    /// Get query for deterministic values
+    fn get_deterministic_samples_query(&self, table_name: String) -> String;
+
+    /// Decodes column to our Content
+    fn decode_to_content(&self, column_info: &ColumnInfo) -> Result<Content>;
+
+    // Returns extended query string + current index
+    fn extend_parameterised_query(query: &mut String, _curr_index: usize, extend: usize) {
+        query.push('(');
+        for i in 0..extend {
+            query.push('?');
```

Comment:
Rather than have `PostgresDataSource` re-implement this function in almost the exact same way, perhaps it could make sense to define a new associate function on the trait that returns either `?` or `$n` depending on whether Postgres or MySQL, and then not override `extend_parameterised_query` in `PostgresDataSource`?

Reply:
Oww, nice idea!!! Maybe I can even get away with an associate type.